src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,378 @@
+    apply rel.neg_add,
+  end,
+  add_comm := λ a b,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    apply quot.sound,
+    apply rel.add_comm,
+  end,
+  mul := λ a b, quot.lift_on a (λ x, quot.lift_on b (λ y, quot.mk _ (x * y))
+  begin
+    intros a b h,
+    dsimp only [],
+    apply quot.sound,
+    apply rel.mul_compat_right h,
+  end)
+  begin
+    intros a b h,
+    dsimp only [],
+    congr,
+    ext,
+    apply quot.sound,
+    apply rel.mul_compat_left h,
+  end,
+  mul_assoc := λ a b c,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    rcases quot.exists_rep c with ⟨c,rfl⟩,
+    apply quot.sound,
+    apply rel.mul_assoc,
+  end,
+  one := quot.mk _ 1,
+  one_mul := λ a,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    apply quot.sound,
+    apply rel.one_mul,
+  end,
+  mul_one := λ a,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    apply quot.sound,
+    apply rel.mul_one,
+  end,
+  left_distrib := λ a b c,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    rcases quot.exists_rep c with ⟨c,rfl⟩,
+    apply quot.sound,
+    apply rel.left_distrib,
+  end,
+  right_distrib := λ a b c,
+  begin
+    rcases quot.exists_rep a with ⟨a,rfl⟩,
+    rcases quot.exists_rep b with ⟨b,rfl⟩,
+    rcases quot.exists_rep c with ⟨c,rfl⟩,
+    apply quot.sound,
+    apply rel.right_distrib,
+  end }
+
+instance : inhabited (tensor_algebra R M) := ⟨0⟩
+
+instance : has_scalar R (tensor_algebra R M) :=
+{ smul := λ r a, quot.lift_on a (λ x, quot.mk _ $ r • x)
+begin
+  intros a b h,
+  dsimp only [],
+  apply quot.sound,
+  apply rel.smul_compat h,
+end }
+
+instance : algebra R (tensor_algebra R M) :=
+{ to_fun := λ r, r • 1,
+  map_one' := by apply quot.sound; apply rel.map_one,
+  map_mul' := λ a b, by apply quot.sound; apply rel.map_mul,
+  map_zero' := by apply quot.sound; apply rel.map_zero,
+  map_add' := λ a b, by apply quot.sound; apply rel.map_add,
+  commutes' := λ r x,
+  begin
+    dsimp only [],
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    apply quot.sound,
+    apply rel.commutes,
+  end,
+  smul_def' := λ r x,
+  begin
+    dsimp only [],
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    apply quot.sound,
+    apply rel.smul_def,
+  end }
+
+/--
+The universal linear map `M →ₗ[R] tensor_algebra R M`.
+-/
+def univ : M →ₗ[R] (tensor_algebra R M) :=
+{ to_fun := λ m, quot.mk _ m,
+  map_add' := λ x y, by apply quot.sound; apply rel.add_lin,
+  map_smul' := λ r x, by apply quot.sound; apply rel.smul_lin }
+
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
+-/
+def lift {A : Type u3} [ring A] [algebra R A] (f : M →ₗ[R] A) : tensor_algebra R M →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
```

Comment:
I think this might be clearer with a `match` expression where you can name each field

Reply:
I'm not sure what you mean. Can you `match` in a tactic block?

Reply:
I think I was mistaken. What you can do is added `cases zero_lin: ` etc before each `{}` block.